### DISCUSSION TITLE
Unwrap function is making the iterations take way too long time. Help!

### DISCUSSION THREAD

**Discussion by FezokaTurbo** [MDAnalysis/mdanalysis#4641] [Q&A] [answered]
Greetings!
I'm currently analysing a MD where one of the monomers touches the border of the water box and gets transported to the other side of the box.
Reading the library, I found the unwrap function, that is supposed to help in these cases. 
The problem lies after the transformation after unwraping is done. Whenever i try to iterate over the Universe's Trajectory, each iteration takes a few seconds. My dynamic has 1500 frames, so it will that some time to even be able to test if my code is written correctly.
Can someone vouch if it is a normal thing to happen?
The code is as follows:

```
atoms = u.select_atoms('protein and name CA and backbone')
transform = trans.unwrap(atoms)
u.trajectory.add_transformations(transform)
print('Unwrap completed.')

#Firstly, we will calculate the Radius of gyration

protein = u.select_atoms('protein')
rgyr = []
time = []
for ts in u.trajectory:
    print(ts) #This point is where I am using to see that the problem lies somewhere around here.
    tempo = (ts.time)/1000
    time.append(tempo)
    rgyr.append(protein.radius_of_gyration())

print(f'Radius of Gyration info was gathered.')
```

Sorry for my english, it isn't my first language. 
Thank you for the time!

**Comment by orbeckst** [comment]
The unwrap *transformation* is applied *every time you go to a new trajectory frame* and unwrap is currently a slow operation. If you go back to a frame, the data are again read from disk, and the unwrap transformation is applied. If you need to repeatedly analyze your trajectory I would save an unwrapped one. This will take time, but only once
```python
u = mda.Universe(TOPOL, "traj.xtc")
atoms = u.select_atoms('protein and name CA and backbone')
transform = trans.unwrap(atoms)
u.trajectory.add_transformations(transform)
u.atoms.write("unwrapped.xtc", frames="all")
```
Then load `unwrapped.xtc`.

If you have other tools at hand that can do unwrap faster (e.g. GROMACS `gmx trjconv`) then I'd use those to create unwrapped trajectories.

**Reply by jochuan** [reply]
Is there a possibility of using unwrap in only the frames where the protein exits from the water box or do i need to use it in all frames of the trajectory for analysis such as RMSD, RMSF, ramachandran and so on? In my case the trajectory has 150ns and it is divided in two parts but when i tried to join these two parts before in only a single file the time of the whole trajectory is reduced:
![image](https://github.com/user-attachments/assets/a54e040d-8789-42db-acb4-e8d8cb899313)

**Reply by hmacdope** [reply]
You need to use it in all frames, as you need to check if anything exits the box at the timestep. IE without checking the unwrapping there is no way to "check" whether something is exiting.

**Reply by jochuan** [reply]
> You need to use it in all frames, as you need to check if anything exits the box at the timestep. IE without checking the unwrapping there is no way to "check" whether something is exiting.

I know the exact frames where the protein is exiting because it is causing one of the monomers to teleport to the other side of the box when looking on the trajectory using VMD

**Reply by FezokaTurbo** [reply, marked as answer]
`if unwrap_switch == 'y':

   transform = [trans.NoJump(protein),
               trans.center_in_box(protein, wrap= True)]
   u.trajectory.add_transformations(*transform)
   print('Transformation preparated.')`

The `transformation.NoJump` was able to resolve our issue. Thank you for the discussion, have a great day!
